File: 01_hello_usdz/main.cpp
// MIT License
// 
// Copyright (C) 2018-2023, Tellusim Technologies Inc. https://tellusim.com/
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <TellusimApp.h>
#include <core/TellusimLog.h>
#include <core/TellusimBlob.h>
#include <core/TellusimTime.h>
#include <format/TellusimMesh.h>
#include <graphics/TellusimMeshModel.h>
#include <platform/TellusimWindow.h>
#include <platform/TellusimDevice.h>
#include <platform/TellusimPipeline.h>
#include <platform/TellusimCommand.h>

/*
 */
using namespace Tellusim;

/*
 */
static Texture create_texture(const Device &device, const MeshMaterial &material, const char *type) {
	
	// fine material parameter
	uint32_t index = material.findParameter(type);
	if(index == Maxu32 || !material.hasParameterFlag(index, MeshMaterial::FlagBlob)) return Texture::null;
	
	// load image
	Image image;
	Blob blob = material.getParameterBlob(index);
	if(!image.load(blob)) return Texture::null;
	
	// create texture
	return device.createTexture(image, Texture::FlagMipmaps);
}

/*
 */
int32_t main(int32_t argc, char **argv) {
	
	// create app
	App app(argc, argv);
	if(!app.create()) return 1;
	
	// create window
	Window window(app.getPlatform(), app.getDevice());
	if(!window || !window.setSize(app.getWidth(), app.getHeight())) return 1;
	if(!window.create("01 Hello USDZ") || !window.setHidden(false)) return 1;
	window.setKeyboardPressedCallback([&](uint32_t key, uint32_t code) {
		if(key == Window::KeyEsc) window.stop();
	});
	
	// create device
	Device device(window);
	if(!device) return 1;
	
	// vertex layout
	struct Vertex {
		Vector3f position;
		Vector3f normal;
		Vector4f tangent;
		Vector2f texcoord;
		Vector4f weights;
		Vector4u joints;
	};
	
	// common parameters
	struct CommonParameters {
		Matrix4x4f projection;
		Matrix4x4f modelview;
		Vector4f camera;
	};
	
	// create pipeline
	Pipeline pipeline = device.createPipeline();
	pipeline.setSamplerMask(0, Shader::MaskFragment);
	pipeline.setTextureMasks(0, 3, Shader::MaskFragment);
	pipeline.setUniformMasks(0, 2, Shader::MaskVertex);
	pipeline.addAttribute(Pipeline::AttributePosition, FormatRGBf32, 0, offsetof(Vertex, position), sizeof(Vertex));
	pipeline.addAttribute(Pipeline::AttributeNormal, FormatRGBf32, 0, offsetof(Vertex, normal), sizeof(Vertex));
	pipeline.addAttribute(Pipeline::AttributeTangent, FormatRGBAf32, 0, offsetof(Vertex, tangent), sizeof(Vertex));
	pipeline.addAttribute(Pipeline::AttributeTexCoord, FormatRGf32, 0, offsetof(Vertex, texcoord), sizeof(Vertex));
	pipeline.addAttribute(Pipeline::AttributeWeights, FormatRGBAf32, 0, offsetof(Vertex, weights), sizeof(Vertex));
	pipeline.addAttribute(Pipeline::AttributeJoints, FormatRGBAu32, 0, offsetof(Vertex, joints), sizeof(Vertex));
	pipeline.setColorFormat(window.getColorFormat());
	pipeline.setDepthFormat(window.getDepthFormat());
	pipeline.setDepthFunc(Pipeline::DepthFuncLess);
	if(!pipeline.loadShaderGLSL(Shader::TypeVertex, "main.shader", "VERTEX_SHADER=1")) return 1;
	if(!pipeline.loadShaderGLSL(Shader::TypeFragment, "main.shader", "FRAGMENT_SHADER=1")) return 1;
	if(!pipeline.create()) return 1;
	
	// load mesh
	Mesh mesh;
	if(!mesh.load("model.usdz")) return 1;
	if(!mesh.getNumGeometries()) return 1;
	if(!mesh.getNumAnimations()) return 1;
	mesh.setBasis(Mesh::BasisZUpRight);
	mesh.createTangents();
	
	// create model
	MeshModel model;
	if(!model.create(device, pipeline, mesh)) return 1;
	
	// create textures
	Array<Texture> normal_textures;
	Array<Texture> diffuse_textures;
	Array<Texture> roughness_textures;
	for(const MeshGeometry &geometry : mesh.getGeometries()) {
		for(const MeshMaterial &material : geometry.getMaterials()) {
			normal_textures.append(create_texture(device, material, MeshMaterial::TypeNormal));
			diffuse_textures.append(create_texture(device, material, MeshMaterial::TypeDiffuse));
			roughness_textures.append(create_texture(device, material, MeshMaterial::TypeRoughness));
		}
	}
	
	// create sampler
	Sampler sampler = device.createSampler(Sampler::FilterTrilinear, Sampler::WrapModeRepeat);
	if(!sampler) return 1;
	
	// create target
	Target target = device.createTarget(window);
	target.setClearColor(Color::gray);
	
	// main loop
	window.run([&]() -> bool {
		
		Window::update();
		
		// render window
		if(!window.render()) return false;
		
		// window target
		target.begin();
		{
			// create command list
			Command command = device.createCommand(target);
			
			// set pipeline
			command.setPipeline(pipeline);
			
			// set sampler
			command.setSampler(0, sampler);
			
			// set model buffers
			model.setBuffers(command);
			
			// set common parameters
			CommonParameters common_parameters;
			common_parameters.camera = Vector4f(0.0f, -180.0f, 180.0f, 0.0f);
			common_parameters.projection = Matrix4x4f::perspective(60.0f, (float32_t)window.getWidth() / window.getHeight(), 0.1f, 1000.0f);
			common_parameters.modelview = Matrix4x4f::lookAt(common_parameters.camera.xyz, Vector3f(0.0f, 0.0f, 80.0f), Vector3f(0.0f, 0.0f, 1.0f));
			if(target.isFlipped()) common_parameters.projection = Matrix4x4f::scale(1.0f, -1.0f, 1.0f) * common_parameters.projection;
			command.setUniform(0, common_parameters);
			
			// mesh animation
			float64_t time = Time::seconds();
			MeshAnimation animation = mesh.getAnimation(0);
			animation.setTime(time, Matrix4x3d::rotateZ(Tellusim::sin(time) * 30.0));
			
			// draw geometries
			uint32_t texture_index = 0;
			Vector4f joint_parameters[192];
			for(const MeshGeometry &geometry : mesh.getGeometries()) {
				
				// joint transforms
				for(uint32_t i = 0, j = 0; i < geometry.getNumJoints(); i++, j += 3) {
					const MeshJoint &joint = geometry.getJoint(i);
					Matrix4x3f transform = Matrix4x3f(animation.getGlobalTransform(joint)) * joint.getITransform() * geometry.getTransform();
					joint_parameters[j + 0] = transform.row_0;
					joint_parameters[j + 1] = transform.row_1;
					joint_parameters[j + 2] = transform.row_2;
				}
				command.setUniform(1, joint_parameters);
				
				// draw materials
				for(const MeshMaterial &material : geometry.getMaterials()) {
					command.setTexture(0, normal_textures[texture_index]);
					command.setTexture(1, diffuse_textures[texture_index]);
					command.setTexture(2, roughness_textures[texture_index]);
					model.draw(command, geometry.getIndex(), material.getIndex());
					texture_index++;
				}
			}
		}
		target.end();
		
		// present window
		if(!window.present()) return false;
		
		// check errors
		if(!device.check()) return false;
		
		return true;
	});
	
	// finish context
	window.finish();
	
	return 0;
}

File: 01_hello_usdz/main.py
#!/usr/bin/env python3

# MIT License
# 
# Copyright (C) 2018-2024, Tellusim Technologies Inc. https://tellusim.com/
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import math

from tellusimd import *

#
# create material texture
#
def create_texture(device, material, type):
	
	# find material parameter
	index = material.findParameter(type)
	if index == Maxu32 or not material.hasParameterFlag(index, MeshMaterial.FlagBlob): return None
	
	# load image
	image = Image()
	blob = material.getParameterBlob(index)
	if not image.load(blob): return None
	
	# create texture
	return device.createTexture(image, Texture.FlagMipmaps)

#
# main
#
def main(argv):
	
	# create app
	app = App(argv)
	if not app.create(): return 1
	
	# create window
	window = Window(app.getPlatform(), app.getDevice())
	if not window or not window.setSize(app.getWidth(), app.getHeight()): return 1
	if not window.create('01 Hello USDZ Python') or not window.setHidden(False): return 1
	window.setKeyboardPressedCallback(lambda key, code: window.stop() if key == Window.KeyEsc else None)
	
	# create device
	device = Device(window)
	if not device: return 1
	
	# create pipeline
	pipeline = device.createPipeline()
	pipeline.setSamplerMask(0, Shader.MaskFragment)
	pipeline.setTextureMasks(0, 3, Shader.MaskFragment)
	pipeline.setUniformMasks(0, 2, Shader.MaskVertex)
	pipeline.addAttribute(Pipeline.AttributePosition, FormatRGBf32, 0, offset = 0, stride = 80)
	pipeline.addAttribute(Pipeline.AttributeNormal, FormatRGBf32, 0, offset = 12, stride = 80)
	pipeline.addAttribute(Pipeline.AttributeTangent, FormatRGBAf32, 0, offset = 24, stride = 80)
	pipeline.addAttribute(Pipeline.AttributeTexCoord, FormatRGf32, 0, offset = 40, stride = 80)
	pipeline.addAttribute(Pipeline.AttributeWeights, FormatRGBAf32, 1, offset = 48, stride = 80)
	pipeline.addAttribute(Pipeline.AttributeJoints, FormatRGBAu32, 1, offset = 64, stride = 80)
	pipeline.setColorFormat(window.getColorFormat())
	pipeline.setDepthFormat(window.getDepthFormat())
	pipeline.setDepthFunc(Pipeline.DepthFuncLess)
	if not pipeline.loadShaderGLSL(Shader.TypeVertex, 'main.shader', 'VERTEX_SHADER=1'): return 1
	if not pipeline.loadShaderGLSL(Shader.TypeFragment, 'main.shader', 'FRAGMENT_SHADER=1'): return 1
	if not pipeline.create(): return 1
	
	# load mesh
	mesh = Mesh()
	if not mesh.load('model.usdz'): return 1
	if not mesh.getNumGeometries(): return 1
	if not mesh.getNumAnimations(): return 1
	mesh.setBasis(Mesh.BasisZUpRight)
	mesh.createTangents()
	
	# create model
	model = MeshModel()
	if not model.create(device, pipeline, mesh): return 1
	
	# create textures
	normal_textures = []
	diffuse_textures = []
	roughness_textures = []
	for geometry in mesh.getGeometries():
		for material in geometry.getMaterials():
			normal_textures.append(create_texture(device, material, MeshMaterial.TypeNormal))
			diffuse_textures.append(create_texture(device, material, MeshMaterial.TypeDiffuse))
			roughness_textures.append(create_texture(device, material, MeshMaterial.TypeRoughness))
	
	# create sampler
	sampler = device.createSampler(Sampler.FilterTrilinear, Sampler.WrapModeRepeat)
	if not sampler: return 1
	
	# create target
	target = device.createTarget(window)
	target.setClearColor(Color.gray)
	
	# main loop
	def main_loop():
		
		Window.update()
		
		if not window.render(): return False
		
		# window target
		if target.begin():
			
			# create command list
			command = device.createCommand(target)
			
			# set pipeline
			command.setPipeline(pipeline)
			
			# set sampler
			command.setSampler(0, sampler)
			
			# set model buffers
			model.setBuffers(command)
			
			# set common parameters
			camera = Vector4f(0.0, -180.0, 180.0, 0.0)
			projection = Matrix4x4f.perspective(60.0, window.getWidth() / window.getHeight(), 0.1, 1000.0)
			modelview = Matrix4x4f.lookAt(camera.xyz, Vector3f(0.0, 0.0, 80.0), Vector3f(0.0, 0.0, 1.0))
			if target.isFlipped(): projection = Matrix4x4f.scale(1.0, -1.0, 1.0) * projection
			common_parameters = bytearray(projection)
			common_parameters += modelview
			common_parameters += camera
			command.setUniform(0, common_parameters)
			
			# mesh animation
			time = Time.seconds()
			animation = mesh.getAnimation(0)
			animation.setTime(time, Matrix4x3d.rotateZ(math.sin(time) * 30.0))
			
			# draw geometries
			texture_index = 0
			for geometry in mesh.getGeometries():
				
				# joint transforms
				joint_parameters = bytearray()
				for joint in geometry.getJoints():
					transform = Matrix4x3f(animation.getGlobalTransform(joint)) * joint.getITransform() * geometry.getTransform()
					joint_parameters += transform.row_0
					joint_parameters += transform.row_1
					joint_parameters += transform.row_2
				command.setUniform(1, joint_parameters)
				
				# draw materials
				for material in geometry.getMaterials():
					command.setTexture(0, normal_textures[texture_index])
					command.setTexture(1, diffuse_textures[texture_index])
					command.setTexture(2, roughness_textures[texture_index])
					model.draw(command, geometry.getIndex(), material.getIndex())
					texture_index += 1
			
			command = None
			
			target.end()
		
		if not window.present(): return False
		
		if not device.check(): return False
		
		return True
	
	window.run(main_loop)
	
	# finish context
	window.finish()
	
	# done
	Log.print('Done\n')
	
	return 0

#
# entry point
#
if __name__ == '__main__':
	try:
		exit(main(sys.argv))
	except Exception as error:
		print('\n' + str(error))
		exit(1)
